app/controllers/i18n/TaiLanguageController.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.i18n
+
+import controllers.auth.WithAuthorisedForTaiLite
+import controllers.{AuthenticationConnectors, ServiceCheckLite, TaiBaseController}
+import play.api.i18n.Lang
+import play.api.mvc.{Action, AnyContent}
+import uk.gov.hmrc.play.frontend.auth.DelegationAwareActions
+import uk.gov.hmrc.play.language.{LanguageController, LanguageUtils}
+import uk.gov.hmrc.play.partials.PartialRetriever
+import uk.gov.hmrc.renderer.TemplateRenderer
+import uk.gov.hmrc.tai.config.{FeatureTogglesConfig, TaiHtmlPartialRetriever}
+import uk.gov.hmrc.tai.connectors.LocalTemplateRenderer
+import uk.gov.hmrc.tai.service.TaiService
+import play.api.Play.current
+import play.api.i18n.Messages
+import play.api.i18n.Messages.Implicits._
+
+import scala.concurrent.Future
+
+trait TaiLanguageController extends LanguageController with TaiBaseController
+  with DelegationAwareActions
+  with WithAuthorisedForTaiLite
+  with FeatureTogglesConfig {
+
+  def taiService: TaiService
+
+  override protected def languageMap: Map[String, Lang] = Map(
+    "english" -> Lang("en"),
+    "cymraeg" -> Lang("cy")
+  )
+
+  override protected def fallbackURL: String = controllers.routes.WhatDoYouWantToDoController.whatDoYouWantToDoPage.url
+
+  protected def isWelshEnabled = welshLanguageEnabled
+
+  override def switchToLanguage (language: String): Action[AnyContent] = authorisedForTai(taiService).async {
+    implicit user =>
+      implicit taiRoot =>
+        implicit request =>
+          ServiceCheckLite.personDetailsCheck {
+
+            val newLanguage =
+              if (isWelshEnabled)
+                languageMap.getOrElse(language, LanguageUtils.getCurrentLang)
+              else
+                LanguageUtils.getCurrentLang
```

Comment:
Will this allow Welsh to be on and unable to switch to English if we toggle the lang switcher off via the flag while the user is mid-session in Welsh?

Reply:
It would yeah, the user would need to clear cookies to revert.
Since we only have two language options I can instead default to english. It's less scalable if we were ever to introduce a third language - but that's not going to happen any time soon